stackhpc_ipa_hardware_managers/system_bios.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2015 Rackspace, Inc.
+# Copyright 2017 StackHPC Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ironic_python_agent import errors
+from ironic_python_agent import hardware
+from ironic_python_agent import utils
+
+from oslo_concurrency import processutils
+
+from oslo_log import log
+
+LOG = log.getLogger()
+
+
+def _get_property(dmidecode_property):
+    system_property = ''
+    try:
+        system_property, _e = utils.execute(
+            "sudo dmidecode -s {}".format(dmidecode_property),
+            shell=True)
+    except (processutils.ProcessExecutionError, OSError) as e:
+            LOG.warning("Cannot read system property: {}".format(e))
+
+    system_property = system_property.rstrip()
+    LOG.debug('System property: {}'.format(system_property))
+    return system_property
+
+
+def _get_product():
+    return _get_property('system-product-name')
+
+
+def _get_bios():
+    return _get_property('bios-version')
+
+
+def _handle_bios_update(actual_bios_version, expected_bios_version):
+    """Handle system BIOS update stub.
+
+    Future managers may override this to support automatic updates.
+    """
+    raise errors.CleaningError(
+        "A manual BIOS update is required. Expected version '{0}' but "
+        "version '{1}' was installed".
+        format(expected_bios_version, actual_bios_version))
+
+
+def _get_expected_property(node, node_property):
+    try:
+        expected_property = node['extra']['system_vendor'][node_property]
+    except KeyError as e:
+        raise errors.CleaningError(
+            "Expected property '{0}' not found. For cleaning to proceed "
+            "you must set the property 'system_vendor/{1}' in the node's "
+            "extra field, for example: $ openstack baremetal node set "
+            "$NODE_ID --extra system_vendor/{1}=$VALUE".format(e.message,
+                                                               node_property))
```

Comment:
There's a part of me that worries about forcing nodes to have this field set, as there isn't an easy way to disable this clean step once it's baked into the ramdisk. I guess it's fine, as long as we're sure all our nodes can pass the step.

Reply:
I could always add another field in ' extra' to make the step a no-op when it's set?

Reply:
I'll leave it up to you.